src/model/consent/consent.ts

```diff
@@ -0,0 +1,141 @@
+ Gates Foundation organization for an example). Those individuals should have
+ their names indented and be marked with a '-'. Email address can be added
+ optionally within square brackets <email>.
+ * Gates Foundation
+ - Name Surname <name.surname@gatesfoundation.com>
+
+ - Raman Mangla <ramanmangla@google.com>
+ --------------
+ ******/
+
+/*
+ * Unlike MySQL, SQLite (testing DB) doesn't support Timestamp type natively.
+ * It returns ISO strings using the inbuilt Date() function.
+ * Thus, typecasting of MySQL Timestamp to Date object for insertion/retrieval
+ * for 'createdAt' field and testing for the same may be required in future.
+ */
+
+import { NotFoundError } from '../errors'
+import Knex from 'knex'
+
+/*
+ * Interface for Consent resource type
+ */
+export interface Consent {
+  id: string;
+  initiatorId: string;
+  participantId: string;
+  credentialId?: string;
+  credentialType?: string;
+  credentialStatus?: string;
+  credentialPayload?: string;
+  credentialChallenge?: string;
+  createdAt?: string;
+}
+
+/*
+ * Class to abstract Consent DB operations
+ */
+export class ConsentDB {
+  // Knex instance
+  private Db: Knex
+  // Nullable fields which don't allow for explicit null upsert
+  private static nullProtectedProps: Record<string, boolean> = {
+    credentialId: true,
+    credentialType: true,
+    credentialStatus: true,
+    credentialPayload: true,
+    credentialChallenge: true
+  }
+
+  public constructor (dbInstance: Knex) {
+    this.Db = dbInstance
+  }
+
+  // Add initial Consent parameters
+  public async register (consent: Consent): Promise<number> {
+    // Returns array containing number of inserted rows
+    const insertCount: number[] = await this
+      .Db<Consent>('Consent')
+      .insert(consent)
+
+    return insertCount[0]
+  }
+
+  // Update Consent credential
+  // Only non-null Consent fields are updated
+  public async updateCredentials (consent: Consent): Promise<number> {
+    const validatedConsent = {}
+
+    for (const key in consent) {
+      // Only allow nullProtectedFields with non-null values for update
+      if (ConsentDB.nullProtectedProps[key] && (consent as unknown as Record<string, boolean>)[key]) {
+        (validatedConsent as Record<string, boolean>)[key] = (consent as unknown as Record<string, boolean>)[key]
+      }
+    }
+
+    // Returns number of updated rows
+    const updateCount: number = await this
+      .Db<Consent>('Consent')
+      .where({ id: consent.id })
+      .update(validatedConsent)
+
+    // Ensure that the caller knows that the resource does not exist
+    if (updateCount === 0) {
+      throw new NotFoundError('Consent', consent.id)
+    }
+
+    return updateCount
+  }
+
+  // Retrieve Consent by ID (unique)
+  public async retrieve (id: string): Promise<Consent[]> {
+    // Returns array containing consents
+    const consents: Consent[] = await this
+      .Db<Consent>('Consent')
+      .select('*')
+      .where({ id: id })
+      .limit(1)
```

Comment:
maybe i'm just misunderstanding but 1) if ID is unique then then why are you returning an array? 2) if you're returning an array what does the `limit(1)` do?